pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -555,12 +575,27 @@ public SegmentMetadata getSegmentMetadata() {
   @Override
   public ColumnDataSource getDataSource(String columnName) {
     FieldSpec fieldSpec = _schema.getFieldSpecFor(columnName);
-    if (fieldSpec.isVirtualColumn()) {
-      VirtualColumnContext virtualColumnContext = new VirtualColumnContext(fieldSpec, _numDocsIndexed);
-      VirtualColumnProvider virtualColumnProvider =
-          VirtualColumnProviderFactory.buildProvider(_schema.getFieldSpecFor(columnName).getVirtualColumnProvider());
-      return new ColumnDataSource(virtualColumnProvider.buildColumnIndexContainer(virtualColumnContext),
-          virtualColumnProvider.buildMetadata(virtualColumnContext));
+    if (fieldSpec == null || fieldSpec.isVirtualColumn()) {
+      // Column is either added during ingestion, or was initiated with a virtual column provider
+      if (fieldSpec == null) {
+        // If the column was added during ingestion, we will construct the virtual column provider based on its fieldSpec
+        fieldSpec = _newlyAddedColumnsFieldMap.get(columnName);
+        Preconditions.checkNotNull(fieldSpec, "FieldSpec for " + columnName + " should not be null");
+      }
+      ColumnContext columnContext = new ColumnContext(fieldSpec, _numDocsIndexed);
+      ColumnProvider columnProvider = _newlyAddedColumnsProviderMap.getOrDefault(columnName, ColumnProviderFactory.buildProvider(columnContext));
+      if (columnProvider instanceof DefaultNullValueColumnProvider) {
```

Comment:
Can we do theese things inside the Column ProviderFFactory so that we dont need to cast here?

Reply:
updated a bit. kept one casting for `updateInvertedIndex`, since now we would like to cache the provider

Reply:
Why cache it? We construct it afresh for each query in the offline case, this should not be worse than that, right?